raysect/core/math/quaternion.pyx

```diff
@@ -0,0 +1,556 @@
+        :return: A new instance of this vector that has been transformed with the supplied Affine Matrix.
+        :rtype: Vector3D
+
+        .. code-block:: pycon
+
+            >>> z = Vector3D(0, 0, 1)
+            >>> y = z.transform(rotate_x(90))
+            >>> y
+            Vector3D(0.0, -1.0, 6.123233995736766e-17)
+        """
+
+        cdef:
+            double qs, qx, qy, qz
+            double qs2, qx2, qy2, qz2
+            double m00, m01, m02, m03
+            double m10, m11, m12, m13
+            double m20, m21, m22, m23
+            double m30, m31, m32, m33
+            double temp1, temp2
+            Quaternion unit_q
+
+        unit_q = self.normalise()
+
+        qs = unit_q.s
+        qx = unit_q.x
+        qy = unit_q.y
+        qz = unit_q.z
+
+        qs2 = qs * qs
+        qx2 = qx * qx
+        qy2 = qy * qy
+        qz2 = qz * qz
+
+        m00 = 1 - 2*qy2 - 2*qz2
+        m01 = 2*qx*qy - 2*qz*qs
+        m02 = 2*qx*qz + 2*qy*qs
+
+        m10 = 2*qx*qy + 2*qz*qs
+        m11 = 1 - 2*qx2 - 2*qz2
+        m12 = 2*qy*qz - 2*qx*qs
+
+        m20 = 2*qx*qz - 2*qy*qs
+        m21 = 2*qy*qz + 2*qx*qs
+        m22 = 1 - 2*qx2 - 2*qy2
+
+        m03 = m13 = m23 = 0.0
+        m30 = m31 = m32 = 0.0
+        m33 = 1.0
+
+        return new_affinematrix3d(m00, m01, m02, m03,
+                                  m10, m11, m12, m13,
+                                  m20, m21, m22, m23,
+                                  m30, m31, m32, m33)
+
+    @classmethod
+    def from_axis_angle(cls, axis, angle):
+        """
+        Generates a new Quaternion from the axis-angle specification.
+
+        :param Vector3D axis: The axis about which rotation will be performed.
+        :param float angle: An angle in degrees specifiying the magnitude of the
+          rotation about the axis vector.
+        :return: A new Quaternion object representing the specified rotation.
+
+        .. code-block:: pycon
+
+           >>> from raysect.core.math import Quaternion, Vector3D
+           >>>
+           >>> Quaternion.from_axis_angle(Vector3D(1, 0, 0), 45)
+           Quaternion(0.9238795325112867, <0.3826834323650898, 0.0, 0.0>)
+        """
+
+        if not isinstance(axis, Vector3D):
+            raise TypeError("The rotation axis must be specified with a Vector3D.")
+
+        if not isinstance(angle, (float, int)):
+            raise TypeError("The rotation angle must be specified with a real number.")
+
+        if not -180 <= angle <= 180:
+            raise ValueError("The angle of rotation must be in the range (-180, 180).")
+
+        axis = axis.normalise()
+
+        theta_2 = angle * DEG2RAD / 2
+
+        qs = cos(theta_2)
+        qx = axis.x * sin(theta_2)
+        qy = axis.y * sin(theta_2)
+        qz = axis.z * sin(theta_2)
+
+        return new_quaternion(qs, qx, qy, qz)
+
+
+cpdef Quaternion matrix2quaternion(AffineMatrix3D matrix):
```

Comment:
I hate this function name! Will think of an alternative.... probably just to_quarternion().

Alternately make this a class method on Quarternion eg: Quarternion.from_matrix() which has nice symmetry with to_matrix.

Reply:
Resolved in a4d9d8a